src/Compiler/TypedTree/TypedTreeOps.fs

```diff
@@ -10108,14 +10148,733 @@ let (|CompiledForEachExpr|_|) g expr =
+
+                    // <body>
+                    // loopVar <- loopVar + step
+                    // i <- i + 1
+                    let body = mkSequentials g mBody [mkBody idxVar loopVar; incrV; incrI]
+
+                    // i < count
+                    let guard = mkAsmExpr ([AI_clt_un], [], [idxVar; count], [g.bool_ty], mFor)
+
+                    // while i < count do
+                    //     <body>
+                    //     loopVar <- loopVar + step
+                    //     i <- i + 1
+                    mkWhile
+                        g
+                        (
+                            spInWhile,
+                            WhileLoopForCompiledForEachExprMarker,
+                            guard,
+                            body,
+                            mBody
+                        )
+                )
+            )
+
+        /// Start at 0 and count up till we have wrapped around.
+        ///
+        ///     <body>
+        ///     while i > 0 do
+        ///         <body>
+        ///         i <- i + 1
+        let mkCountUpInclusive mkBody countTy =
+            mkCompGenLetMutableIn mIn "i" countTy (mkTypedZero g mIn countTy) (fun (idxVal, idxVar) ->
+                mkCompGenLetMutableIn mIn "loopVar" rangeTy start (fun (loopVal, loopVar) ->
+                    // loopVar <- loopVar + step
+                    let incrV = mkValSet mIn (mkLocalValRef loopVal) (mkAsmExpr ([AI_add], [], [loopVar; step], [rangeTy], mIn))
+
+                    // i <- i + 1
+                    let incrI = mkValSet mIn (mkLocalValRef idxVal) (mkAsmExpr ([AI_add], [], [idxVar; mkTypedOne g mIn countTy], [rangeTy], mIn))
+
+                    // <body>
+                    // loopVar <- loopVar + step
+                    // i <- i + 1
+                    let body = mkSequentials g mBody [mkBody idxVar loopVar; incrV; incrI]
+
+                    // i > 0
+                    let guard = mkAsmExpr ([AI_cgt_un], [], [idxVar; mkTypedZero g mFor countTy], [g.bool_ty], mFor)
+
+                    // while i > 0 do
+                    //     <body>
+                    //     loopVar <- loopVar + step
+                    //     i <- i + 1
+                    let loop =
+                        mkWhile
+                            g
+                            (
+                                spInWhile,
+                                WhileLoopForCompiledForEachExprMarker,
+                                guard,
+                                body,
+                                mBody
+                            )
+
+                    // <body>
+                    // loopVar <- loopVar + step
+                    // i <- i + 1
+                    // while i > 0 do
+                    //     <body>
+                    //     loopVar <- loopVar + step
+                    //     i <- i + 1
+                    mkSequential mBody body loop
```

Comment:
(This isn't a huge problem[^1], and it can always be improved later, but the fact that it isn't as good as it could be bothers me.)

As it is, we're actually emitting the instructions for `body` twice here — one copy before the loop and one copy inside it. This works, but it increases the emitted code size, especially if `body` is large.

It would be better if we could emit the equivalent of

```csharp
while (true)
{
    body(i, loopVar);
    loopVar += step;
    i++;
    if (i == 0UL) break;
}
```

The tricky part is that:

- On the one hand, when this function is used for `for`-loops, its output still gets passed through the optimizer, which means there can't already be any `TOp.Label`s or `TOp.Goto`s.
- On the other hand, when this function is used for computed collections, its output _doesn't_ get passed through the optimizer, which means that whatever we emit needs to already be fairly low-level.

Anybody have any ideas for how to emit a proper `do`-`while`-loop here without using `TOp.Label` or `TOp.Goto`?

[^1]: This code will only be emitted if the type being iterated over is `int64`/`uint64`/`nativeint`/`unativeint` and `step` is not known at compile-time, and it will only actually be executed if $step = 1$ and $|finish - start| = 2^{64} + 1$.

Reply:
I mean, I guess we could just add another variable and emit something like

```fsharp
let mutable guard = true
let mutable loopVar = start
let mutable i = 0UL
while guard do
    body i loopVar
    loopVar <- loopVar + step
    i <- i + 1UL
    guard <- i <> 0UL
```

Not pretty, but maybe it's better than nothing?

Reply:
I went ahead and did this in e79ac50.

Reply:
I guess we could also consider just emitting different code depending on whether we're initializing an array or doing anything else, since we know that we can never actually create an array of length $2^{64} + 1$, and so we never actually need to emit this code in that case even if we're initializing an array from a range of 64-bit integers...

But I think I'd probably rather leave that to a followup PR.